h2c/h2c.go

```diff
@@ -57,20 +57,20 @@ func (s Server) Serve(l net.Listener) error {
 			if http2VerboseLogs {
 				log.Debugf("Attempting h2c with prior knowledge.")
 			}
-			conn, err := initH2CWithPriorKnowledge(w)
+			conn, err := initH2CWithPriorKnowledge(w, s.closeHijackedConnQuietly)
 			if err != nil {
 				if http2VerboseLogs {
 					log.Debugf("Error h2c with prior knowledge: %v", err)
 				}
 				return
 			}
-			defer conn.Close()
+			defer s.closeHijackedConnQuietly(conn)
 			h2cSrv := &http2.Server{}
 			h2cSrv.ServeConn(conn, &http2.ServeConnOpts{Handler: originalHandler})
 			return
 		}
-		if conn, err := h2cUpgrade(w, r); err == nil {
-			defer conn.Close()
+		if conn, err := h2cUpgrade(w, r, s.closeHijackedConnQuietly); err == nil {
+			defer s.closeHijackedConnQuietly(conn)
```

Comment:
Couldn't we be smarter within closeHijackedConnQuietly, so that here we could systematically call s.closeHijackedConnQuietly regardless of the value of err, and it would be a NOOP when there was an error that occurred before the connection was hijacked?
Then we would not need to call closeHijackedConnQuietly L183, and there would be no need to pass it as an argument.

Reply:
The main reason for passing it as argument is for handling connection closing in cases where the connection already hijacked, but is not returned (both `initH2CWithPriorKnowledge` and `h2cUpgrade` have such cases).

Another option would be to return the hijacked connection anyway even if an error occurred, but that would mean passing responsibility of cleanup to the caller.